CUDA Graph Compute Function Refactor (precursor for performance improvements)

aendk opened the pull request:
Hi All,
I am working on improving llama.cpp's CUDA graph performance on behalf of NVIDIA.
In preliminary testing, we are seeing up to 3% of performance gain by overlapping CPU and GPU work, and by improving CPU -> GPU copy scheduling on a high end system. The changes are likely to be even more impactful on less capable hardware.

To pave the way for these changes (and to provide readable diffs), I first isolated the cosmetic changes in this PR.
 This PR does **not** contain any changes in the logic. It merely slims down the `ggml_backend_cuda_graph_compute()` by moving certain loops and other subtasks of the original function into 5 new functions.

These changes considerably improve the readability and future maintainability of this part of the CUDA backend.


Should I add prefixes to the new function names, and if so, what do you suggest?

@agray3 @mtavenrath

aendk committed:
Refactor: Moves cuda graph executable update step to separate function.

aendk commented:
FYI: setting Status to Draft whilst I investigate the failed tests.

ggerganov commented:
> FYI: setting Status to Draft whilst I investigate the failed tests.

I think you just need to make the functions `static`.

aendk committed:
Refactor: Moves cuda graph update check to separate function.

aendk committed:
Refactor: Moves cuda graph maintenance (update or adjusting copy parameters) to separate function for improved readability.

aendk committed:
Fix: Adds missing reference to maintain_cuda_graph() definition.

aendk committed:
Refactor: Improves structure and abstractions by moving CUDA graph evaluation and capture to its own function.

aendk committed:
Refactor: Moves node graph checks and copy ops into individual function for improved readability.

aendk committed:
Refactor: Removes code permanently excluded from compilation to increase readability.

aendk committed:
Style: Adds missing newline

aendk commented:
Ok, third time's a charm. My first try lacked the `static` keyword, my second showed that there were some issues around missing `ifdef`'s when disabling CUDA graphs.
 
Now this refactor-only PR is ready for review. Due to the `ifdef`s, the overall diff looks quite messy.

I strongly suggest to review commit by commit. Its easier to see there that not much has happened. 
I moved some code into 5 new functions and removed a few lines of unused/dead code.

slaren commented on ggml/src/ggml-cuda/ggml-cuda.cu:
```diff
@@ -2337,147 +2401,111 @@ static bool ggml_graph_node_has_matching_properties(ggml_tensor * node, ggml_gra
-#endif
-        }
+#ifdef USE_CUDA_GRAPH
+static bool is_cuda_graph_update_required(ggml_backend_cuda_context * cuda_ctx, ggml_cgraph * cgraph, bool cuda_graph_update_required) {
```
I don't think that `cuda_graph_update_required` needs to be a parameter here, it should be just a local.

aendk replied:
you are right.

slaren reviewed (commented):
There are multiple consecutive `#ifdef USE_CUDA_GRAPH`, consider consolidating into a single one.

aendk committed:
Style: Consolidates several neighboring '#ifdef USE_CUDA_GRAPH' into a single one

aendk committed:
Refactor: Makes 'cuda_graph_update_required' a local variable

aendk commented:
I implemented your requested changes. The single test failure looks like a CI / runner issue. 
Feel free to re-review or relaunch the failed test.

slaren committed:
remove double lines between functions
